src/pyspiffe/workloadapi/default_workload_api_client.py

```diff
@@ -34,6 +37,51 @@
     WORKLOAD_API_HEADER_VALUE,
 )
 
+# GRPC Error Codes that the client will not retry on:
+#  - INVALID_ARGUMENT is not retried according to the SPIFFE spec because the request is invalid
+#  - CANCELLED is not retried because it occurs when the caller has canceled the operation.
+_NON_RETRYABLE_CODES = {grpc.StatusCode.CANCELLED, grpc.StatusCode.INVALID_ARGUMENT}
+
+
+class _RetryHandler:
+    def __init__(
+        self,
+        max_retries: int = 0,
+        base_backoff_in_seconds: float = 0.1,
+        backoff_factor: int = 2,
+        max_delay_in_seconds: int = 60,
+    ) -> None:
+        self._max_retries = max_retries
+        self._base_backoff = base_backoff_in_seconds
+        self._backoff_factor = backoff_factor
+        self._max_delay_in_seconds = max_delay_in_seconds
+        self._retries_count = 0
+        self._lock = threading.RLock()
+
+    def do_retry(self, fn: Callable, params: List) -> bool:
+        with self._lock:
+            if self._max_retries and self._retries_count > self._max_retries:
```

Comment:
As _retries_count starts with 0 and _max_retries is increased *after* this test, it should be
the test should be if self._max_retries and self._retries_count >= self._max_retries:
if retries_count = 0 and _max_retries = 1, do_retry could be called 3 times before falling in this if:
first time: retries_count = 0 and _max_retries = 1: self._retries_count > self._max_retries ? False
second time: retries_count = 1 and _max_retries = 1: self._retries_count > self._max_retries ? False
third time: retries_count = 2 and _max_retries = 1: self._retries_count > self._max_retries ? True

Only then, it would stop retries.... so, code would be executed twice instead of once as expected (_max_retries = 1).

Reply:
Good catch!